In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sqlite3
import warnings

warnings.filterwarnings("ignore")
pd.set_option("display.max_columns", None)
pd.set_option('display.float_format', '{:.2f}'.format)

# 1. Data Preparation

## 1.1 Loading and cleaning both the datasets
gas_data = Exports of Goods and Services (WB_WDI_NE_EXP_GNFS_ZS.csv) \
sva_data = Services, Value Added (services_,_value_addes.csv)

In [28]:
gas_data = pd.read_csv("Datasets/WB_WDI_NE_EXP_GNFS_ZS.csv")
sva_data = pd.read_csv("Datasets/services_,_value_addes.csv")

print("Datasets loaded successfully.")
print(f"Services dataset: {sva_data.shape}")
print(f"Exports dataset: {gas_data.shape}")

Datasets loaded successfully.
Services dataset: (10765, 41)
Exports dataset: (11390, 41)


In [62]:
gas_data.sample(5)

,STRUCTURE,STRUCTURE_ID,ACTION,FREQ,REF_AREA,INDICATOR,SEX,AGE,URBANISATION,UNIT_MEASURE,COMP_BREAKDOWN_1,COMP_BREAKDOWN_2,COMP_BREAKDOWN_3,TIME_PERIOD,OBS_VALUE,AGG_METHOD,UNIT_TYPE,DECIMALS,DATABASE_ID,TIME_FORMAT,UNIT_MULT,OBS_STATUS,OBS_CONF,FREQ_LABEL,REF_AREA_LABEL,INDICATOR_LABEL,SEX_LABEL,AGE_LABEL,URBANISATION_LABEL,UNIT_MEASURE_LABEL,COMP_BREAKDOWN_1_LABEL,COMP_BREAKDOWN_2_LABEL,COMP_BREAKDOWN_3_LABEL,AGG_METHOD_LABEL,UNIT_TYPE_LABEL,DECIMALS_LABEL,DATABASE_ID_LABEL,TIME_FORMAT_LABEL,UNIT_MULT_LABEL,OBS_STATUS_LABEL,OBS_CONF_LABEL
1417,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,HPC,WB_WDI_NE_EXP_GNFS_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,1997,20.53,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Heavily indebted poor countries (HIPC),Exports of goods and services (% of GDP),Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public
9631,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,KOR,WB_WDI_NE_EXP_GNFS_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,1982,26.69,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,"Korea, Rep.",Exports of goods and services (% of GDP),Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public
8854,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,TJK,WB_WDI_NE_EXP_GNFS_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,1993,28.55,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Tajikistan,Exports of goods and services (% of GDP),Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public
4535,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,DZA,WB_WDI_NE_EXP_GNFS_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,1984,25.71,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Algeria,Exports of goods and services (% of GDP),Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public
8127,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,PER,WB_WDI_NE_EXP_GNFS_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,2018,25.19,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Peru,Exports of goods and services (% of GDP),Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public


In [63]:
sva_data.sample(5)

,STRUCTURE,STRUCTURE_ID,ACTION,FREQ,REF_AREA,INDICATOR,SEX,AGE,URBANISATION,UNIT_MEASURE,COMP_BREAKDOWN_1,COMP_BREAKDOWN_2,COMP_BREAKDOWN_3,TIME_PERIOD,OBS_VALUE,AGG_METHOD,UNIT_TYPE,DECIMALS,DATABASE_ID,TIME_FORMAT,UNIT_MULT,OBS_STATUS,OBS_CONF,FREQ_LABEL,REF_AREA_LABEL,INDICATOR_LABEL,SEX_LABEL,AGE_LABEL,URBANISATION_LABEL,UNIT_MEASURE_LABEL,COMP_BREAKDOWN_1_LABEL,COMP_BREAKDOWN_2_LABEL,COMP_BREAKDOWN_3_LABEL,AGG_METHOD_LABEL,UNIT_TYPE_LABEL,DECIMALS_LABEL,DATABASE_ID_LABEL,TIME_FORMAT_LABEL,UNIT_MULT_LABEL,OBS_STATUS_LABEL,OBS_CONF_LABEL
7869,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,LAO,WB_WDI_NV_SRV_TOTL_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,2005,43.39,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Lao PDR,"Services, value added (% of GDP)",Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public
9107,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,LUX,WB_WDI_NV_SRV_TOTL_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,2024,81.18,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Luxembourg,"Services, value added (% of GDP)",Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public
5691,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,COL,WB_WDI_NV_SRV_TOTL_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,2018,57.65,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Colombia,"Services, value added (% of GDP)",Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public
2694,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,MEX,WB_WDI_NV_SRV_TOTL_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,2005,60.04,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Mexico,"Services, value added (% of GDP)",Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public
9073,datastructure,WB.DATA360:DS_DATA360(1.3),I,A,BTN,WB_WDI_NV_SRV_TOTL_ZS,_T,_T,_T,PT_GDP,_Z,_Z,_Z,1992,38.57,_Z,RATIO,2,WB_WDI,P1Y,0,A,PU,Annual,Bhutan,"Services, value added (% of GDP)",Total,All age ranges or no breakdown by age,Total,Percentage of GDP,Not Applicable,Not Applicable,Not Applicable,Not Applicable,Ratio,Two,World Development Indicators (WDI),Annual,Units,Normal value,Public


In [29]:
# These are the columns required for analysis
analysis_columns = [
    "REF_AREA",
    "REF_AREA_LABEL",
    "INDICATOR",
    "INDICATOR_LABEL",
    "TIME_PERIOD",
    "OBS_VALUE",
    "UNIT_MEASURE_LABEL",
    "UNIT_TYPE_LABEL"
]

### Choosing the analysis columns

Both datasets contain many columns that repeat metadata about the source, formatting, or reporting structure rather than the actual data. To keep the datasets focused on what matters for analysis, we selected only the columns that describe the country, the indicator being measured, the year, the value itself, and the unit of measurement:

- REF_AREA and REF_AREA_LABEL identify which country the row belongs to
- INDICATOR and INDICATOR_LABEL describe what is being measured
- TIME_PERIOD is the year the value applies to
- OBS_VALUE is the actual figure being analysed
- UNIT_MEASURE_LABEL and UNIT_TYPE_LABEL describe what kind of measurement the value represents (for example, percentage of GDP)

Keeping INDICATOR_LABEL and UNIT_MEASURE_LABEL, even though every remaining row shares the same value, also served a practical purpose, it let us confirm that both datasets were consistently measured in percentage of GDP before comparing them, and it helped catch that the Services Value Added dataset initially contained mixed currencies and growth rates that needed to be filtered out.

In [30]:
# Filtering datasets to South Africa
south_africa_exports = gas_data[gas_data["REF_AREA"] == "ZAF"].copy()
south_africa_services = sva_data[sva_data["REF_AREA"] == "ZAF"].copy()

In [31]:
# Keeping only the relevant analysis columns
exports_analysis = south_africa_exports[analysis_columns].copy()
services_analysis = south_africa_services[analysis_columns].copy()

In [32]:
# Standardising column names
rename_columns = {
    "REF_AREA": "Country_Code",
    "REF_AREA_LABEL": "Country",
    "INDICATOR": "Indicator_Code",
    "INDICATOR_LABEL": "Indicator",
    "TIME_PERIOD": "Year",
    "OBS_VALUE": "Value",
    "UNIT_MEASURE_LABEL": "Unit",
    "UNIT_TYPE_LABEL": "Unit_Type"
}

exports_analysis.rename(columns=rename_columns, inplace=True)
services_analysis.rename(columns=rename_columns, inplace=True)

In [33]:
# Sorting the year chronologically
exports_analysis.sort_values("Year", inplace=True)
services_analysis.sort_values("Year", inplace=True)

In [47]:
# Reseting indexes
exports_analysis.reset_index(drop=True, inplace=True)
services_analysis.reset_index(drop=True, inplace=True)

print("Data preparation completed successfully.")

Data preparation completed successfully.


In [48]:
# Verifying the indicators and measurement units

indicator_check = pd.DataFrame({
    "Dataset": [
        "Exports",
        "Services Value Added"
    ],
    "Country": [
        exports_analysis["Country"].iloc[0],
        services_analysis["Country"].iloc[0]
    ],
    "Indicator": [
        exports_analysis["Indicator"].iloc[0],
        services_analysis["Indicator"].iloc[0]
    ],
    "Unit": [
        exports_analysis["Unit"].iloc[0],
        services_analysis["Unit"].iloc[0]
    ]
})

display(indicator_check)

,Dataset,Country,Indicator,Unit
0,Exports,South Africa,Exports of goods and services (% of GDP),Percentage of GDP
1,Services Value Added,South Africa,"Services, value added (% of GDP)",Percentage of GDP


## 1.2. Data Quality and Validation

Before conducting statistical analysis, the datasets are checked for 
completeness, duplicate country-year observations, temporal coverage and 
consistent measurement units.

These checks ensure that the analysis is based on valid and comparable 
observations.

In [49]:
# Data quality summary

quality_summary = pd.DataFrame({
    "Dataset": [
        "Exports of Goods and Services",
        "Services Value Added"
    ],
    "Rows": [
        len(exports_analysis),
        len(services_analysis)
    ],
    "Columns": [
        exports_analysis.shape[1],
        services_analysis.shape[1]
    ],
    "Missing Values": [
        exports_analysis["Value"].isna().sum(),
        services_analysis["Value"].isna().sum()
    ],
    "Duplicate Country-Year Records": [
        exports_analysis.duplicated(["Country_Code", "Year"]).sum(),
        services_analysis.duplicated(["Country_Code", "Year"]).sum()
    ],
    "Minimum Year": [
        exports_analysis["Year"].min(),
        services_analysis["Year"].min()
    ],
    "Maximum Year": [
        exports_analysis["Year"].max(),
        services_analysis["Year"].max()
    ]
})

display(quality_summary)

,Dataset,Rows,Columns,Missing Values,Duplicate Country-Year Records,Minimum Year,Maximum Year
0,Exports of Goods and Services,66,8,0,0,1960,2025
1,Services Value Added,66,8,0,0,1960,2025


### Data Quality Insight

Both South African datasets contain 66 annual observations covering 1960 to 2025. There are no missing values in the observation field and no duplicate country-year records in either dataset. Both indicators are measured as a percentage of GDP, which makes them directly comparable in the subsequent analysis.

The cleaning process also reduced the original 41-column datasets to eight analysis-relevant variables, removing metadata that is not required for the statistical analysis while retaining the information necessary to identify the country, indicator, year, value and measurement unit.

## 1.3 Descriptive Statistics

Descriptive statistics are used to summarise the distribution and variability
of South Africa's exports of goods and services and services value added as
percentages of GDP.

The analysis considers the mean, median, standard deviation, minimum,
maximum, quartiles and range. Comparing these statistics helps identify the
typical contribution of each indicator to GDP and the extent to which the
indicators have varied between 1960 and 2025.

In [50]:
# Calculating descriptive statistics for both indicators

descriptive_stats = pd.DataFrame({
    "Statistic": [
        "Count",
        "Mean",
        "Median",
        "Standard Deviation",
        "Minimum",
        "25th Percentile",
        "75th Percentile",
        "Maximum",
        "Range"
    ],
    "Exports (% GDP)": [
        exports_analysis["Value"].count(),
        exports_analysis["Value"].mean(),
        exports_analysis["Value"].median(),
        exports_analysis["Value"].std(),
        exports_analysis["Value"].min(),
        exports_analysis["Value"].quantile(0.25),
        exports_analysis["Value"].quantile(0.75),
        exports_analysis["Value"].max(),
        exports_analysis["Value"].max() - exports_analysis["Value"].min()
    ],
    "Services Value Added (% GDP)": [
        services_analysis["Value"].count(),
        services_analysis["Value"].mean(),
        services_analysis["Value"].median(),
        services_analysis["Value"].std(),
        services_analysis["Value"].min(),
        services_analysis["Value"].quantile(0.25),
        services_analysis["Value"].quantile(0.75),
        services_analysis["Value"].max(),
        services_analysis["Value"].max() - services_analysis["Value"].min()
    ]
})

display(descriptive_stats.round(2))

,Statistic,Exports (% GDP),Services Value Added (% GDP)
0,Count,66.00,66.00
1,Mean,25.15,57.48
2,Median,25.09,57.13
3,Standard Deviation,3.81,5.33
4,Minimum,18.40,46.89
5,25th Percentile,22.05,53.03
6,75th Percentile,27.67,62.74
7,Maximum,33.54,64.61
8,Range,15.14,17.72


In [51]:
# Comparing variability

exports_std = exports_analysis["Value"].std()
services_std = services_analysis["Value"].std()

if exports_std > services_std:
    more_variable = "Exports of goods and services"
else:
    more_variable = "Services value added"

print(f"Exports standard deviation: {exports_std:.2f}")
print(f"Services value added standard deviation: {services_std:.2f}")
print(f"More variable indicator: {more_variable}")

Exports standard deviation: 3.81
Services value added standard deviation: 5.33
More variable indicator: Services value added


In [52]:
# Comparing mean and median

exports_mean_median_difference = (
    exports_analysis["Value"].mean() -
    exports_analysis["Value"].median()
)

services_mean_median_difference = (
    services_analysis["Value"].mean() -
    services_analysis["Value"].median()
)

print(
    f"Exports mean-median difference: "
    f"{exports_mean_median_difference:.2f}"
)

print(
    f"Services value added mean-median difference: "
    f"{services_mean_median_difference:.2f}"
)

Exports mean-median difference: 0.06
Services value added mean-median difference: 0.35


In [53]:
# Identifying the highest and lowest observations

extreme_values = pd.DataFrame({
    "Indicator": [
        "Exports of goods and services",
        "Exports of goods and services",
        "Services value added",
        "Services value added"
    ],
    "Observation": [
        "Highest",
        "Lowest",
        "Highest",
        "Lowest"
    ],
    "Year": [
        exports_analysis.loc[exports_analysis["Value"].idxmax(), "Year"],
        exports_analysis.loc[exports_analysis["Value"].idxmin(), "Year"],
        services_analysis.loc[services_analysis["Value"].idxmax(), "Year"],
        services_analysis.loc[services_analysis["Value"].idxmin(), "Year"]
    ],
    "Value (% GDP)": [
        exports_analysis["Value"].max(),
        exports_analysis["Value"].min(),
        services_analysis["Value"].max(),
        services_analysis["Value"].min()
    ]
})

display(extreme_values.round(2))

,Indicator,Observation,Year,Value (% GDP)
0,Exports of goods and services,Highest,2022,33.54
1,Exports of goods and services,Lowest,1970,18.40
2,Services value added,Highest,2020,64.61
3,Services value added,Lowest,1980,46.89


In [54]:
# Calculating the coefficience of variation

exports_cv = (
    exports_analysis["Value"].std() /
    exports_analysis["Value"].mean()
) * 100

services_cv = (
    services_analysis["Value"].std() /
    services_analysis["Value"].mean()
) * 100

cv_comparison = pd.DataFrame({
    "Indicator": [
        "Exports of goods and services",
        "Services value added"
    ],
    "Mean": [
        exports_analysis["Value"].mean(),
        services_analysis["Value"].mean()
    ],
    "Standard Deviation": [
        exports_analysis["Value"].std(),
        services_analysis["Value"].std()
    ],
    "Coefficient of Variation (%)": [
        exports_cv,
        services_cv
    ]
})

display(cv_comparison.round(2))

,Indicator,Mean,Standard Deviation,Coefficient of Variation (%)
0,Exports of goods and services,25.15,3.81,15.14
1,Services value added,57.48,5.33,9.26


### Insight

Services value added averaged 57.48% of GDP, more than double the export figure of 25.15%. This shows services are the dominant contributor to South Africa's economic output, while exports make up a much smaller share.

Looking at variation, services value added has a higher standard deviation (5.33 vs 3.81), so it swings more in absolute terms. But once each indicator is compared against its own average using the coefficient of variation, exports are actually the more volatile of the two (15.14% vs 9.26%). This means exports fluctuate more relative to their own typical size, even though services move more in raw percentage points.

The mean and median are close for both indicators (a difference of 0.06 for exports, 0.35 for services), so neither indicator is being pulled far off-centre by a small number of extreme years, both are fairly evenly distributed across the period.

Exports peaked at 33.54% in 2022 and hit their lowest point at 18.40% in 1970. Services value added peaked at 64.61% in 2020 and was at its lowest, 46.89%, in 1980. The 2020 peak for services lines up with the global shift toward services and remote economic activity during that period, and the 2022 export peak may reflect a post-pandemic trade recovery, both years worth flagging for a closer look in the time-series visualisation.